samples/net/sockets/echo_client/src/udp.c

```diff
@@ -32,7 +32,8 @@ static int send_udp_data(struct data *data)
 
 	do {
 		data->udp.expecting = sys_rand32_get() % ipsum_len;
-	} while (data->udp.expecting == 0U);
+	} while (data->udp.expecting == 0U ||
+		 data->udp.expecting > data->udp.mtu);
```

Comment:
Usually, application should never care about the MTU. A proper fix would be to loop over send() until all the data has been sent (or an error occurred).

Actually, I just checked: look at tcp.c of the same sample, it does exactly that through a local sendall() function.

Reply:
@tbursztyka I dont think UDP should work with the send() loop in the same way as TCP, as most protocols will need protocol header information at the beginning of each packet.  However, for echo_client it would probably work.

The question is: what's the right behavior?

Reply:
@tbursztyka one idea: is to save the results of `send()` back to `data->udp.expecting` (not use a loop).  This would take care of any MTU limit imposed by the network interface.

For reference, here are a few links to existing code in Zephyr for UDP-based samples/libraries:
Coap-Client: https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/coap_client/src/coap-client.c#L187
UpdateHub: https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/updatehub/updatehub.c#L306
LwM2M: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/lwm2m/lwm2m_engine.c#L1031

Reply:
Theoretically you cannot use send() positive result as the MTU as it may send less. On zephyr it will send as the size of MTU because it's a simple stack but your application code should not care about the stack internals.

Btw, same logic applies to recv() (if for instance you request to read less than what has/is going to be received etc...)

If a protocol has internal optimization depending on the MTU, then that's another story (for instance when such protocol manages re-transmission of missed packet instead of re-sending the whole data all over again). But so far I don't think any of the mentioned protocol manages that (maybe I am wrong).

Anyway, here the point of echo application is to send an arbitrary amount of data, no matter the MTU size. Thus why tcp side is already doing it properly btw.

Reply:
I don't think we should consider any kind of UDP fragmentation at the application (socket) level. UDP (DGRAM) sockets behave differently from TCP (STREAM) - you cannot `recv` datagram in chunks and I'd expect the same from `send`. While it's not clearly stated in `send` manual, it's explicit for the `recv`:
```
For message-based sockets, such as SOCK_DGRAM and SOCK_SEQPACKET, the entire message shall be read in
a  single  operation.  If a message is too long to fit in the supplied buffer, and MSG_PEEK is not set in the
flags argument, the excess bytes shall be discarded.
```
For `send` we have:
```
If the message is too long to pass through the underlying protocol, send() shall fail and no data shall
be transmitted.
```
and the following errno code specified:
```
EMSGSIZE
              The message is too large to be sent all at once, as the socket requires.
```

As for the MTU management, I'd expect this to be an IP layer responsibility to fragment the packet at IP level (shouldn't be the case for IPv6 though as the minimal MTU for this protocol is 1280 bytes). But it seems that fragmentation is not handled well as we discuss here.

One possibility to consider to relieve the application from a manual MTU management would be to handle `EMSGSIZE` returned by `send`, and simply re-roll the packet length in that case instead of failing. Up to you to decide what'd be neater.

Reply:
As TCP is a stream protocol, we obviously need to handle partial sends. For UDP, the application normally just sends discrete packets. In order to figure out what size to use for a packet, the MTU is important information. I can also envision a system where the app would try to send large packets and then re-send the data that was not sent. So it really depends on application logic. IMHO using MTU information here is just fine.

Reply:
@rlubos I thought EMSGSIZE applied only to sendmsg since it details after, that for send() no implicit error code is ever returned, just -1. But maybe I am confused by how the man page sentences ore ordered here.

Reply:
@tbursztyka 
Well, I agree these man pages are sometimes hard to follow. If you refer to:
```
If the message is too long to pass atomically through the underlying protocol, the error EMSGSIZE is returned, and the message is not transmitted.

No indication of failure to deliver is implicit in a send(). Locally detected errors are indicated by a return value of -1.
```
which is part of the Linux Programmer's manual (`man 2 send`), I don't think the two paragraphs are related. The second one seems to be formulated cleaner in the POSIX Programmer's manual (`man 3 send`):
```
Successful completion of a call to send() does not guarantee delivery of the message. A return  value  of  −1 indicates only locally-detected errors
```